idx/memory/memory.go

```diff
@@ -465,6 +467,214 @@ func (m *MemoryIdx) TagDetails(orgId int, key, filter string, from int64) (map[s
 	return res, nil
 }
 
+// AutoCompleteTags is used to generate a list of possible values that could
+// complete a given prefix. It also accepts additional conditions to further
+// narrow down the result set.
+//
+// tagPrefix:  the string to be completed
+// expressions: tagdb expressions in the same format as graphite uses
+// from:        only tags will be returned that have at least one metric
+//              with a LastUpdateTime above from
+// limit:       the maximum number of results to return, the results will
+//              always be sorted alphabetically for consistency between
+//              consecutive queries
+//
+func (m *MemoryIdx) AutoCompleteTags(orgId int, tagPrefix string, expressions []string, from int64, limit uint16) ([]string, error) {
+	res := make([]string, 0)
+
+	// only if expressions are specified we need to build a tag query.
+	// otherwise, the generation of the result set is much simpler
+	if len(expressions) > 0 {
+		if len(tagPrefix) > 0 {
+			expressions = append(expressions, fmt.Sprintf("__tag^=%s", tagPrefix))
+		}
+		query, err := NewTagQuery(expressions, from)
+		if err != nil {
+			return nil, err
+		}
+
+		// only acquire lock after we're sure the query is valid
+		m.RLock()
+		defer m.RUnlock()
+
+		tags, ok := m.tags[orgId]
+		if !ok {
+			return nil, nil
+		}
+
+		resMap := query.RunGetTags(tags, m.DefById)
+		for tag := range resMap {
+			res = append(res, tag)
+		}
+
+		// special case for the tag "name", we need to fake it
+		if len(res) > 0 && (len(tagPrefix) >= 4 && tagPrefix[:4] == "name") || len(tagPrefix) == 0 {
+			res = append(res, "name")
+		}
+
+		sort.Strings(res)
+		if uint16(len(res)) > limit && limit > 0 {
+			res = res[:limit]
+		}
+	} else {
+		m.RLock()
+		defer m.RUnlock()
+
+		tags, ok := m.tags[orgId]
+		if !ok {
+			return nil, nil
+		}
+
+		tagsSorted := make([]string, 0, len(tags))
+		for tag := range tags {
+			if len(tagPrefix) > 0 && (len(tagPrefix) > len(tag) || tag[:len(tagPrefix)] != tagPrefix) {
+				continue
+			}
+
+			tagsSorted = append(tagsSorted, tag)
+		}
+
+		sort.Strings(tagsSorted)
+
+	TAGS:
+		for _, tag := range tagsSorted {
+			for _, ids := range tags[tag] {
+				for id := range ids {
+					// only if from is specified we need to find at least one
+					// metric with LastUpdateTime >= from
+					if from > 0 {
+						def, ok := m.DefById[id.String()]
+						if !ok {
+							corruptIndex.Inc()
+							log.Error(3, "memory-idx: corrupt. ID %q is in tag index but not in the byId lookup table", id.String())
+							continue
+						}
+						if def.LastUpdate < from {
+							continue
+						}
+					}
+					res = append(res, tag)
+					if uint16(len(res)) >= limit {
```

Comment:
Seems like this can return a single result for a limit of 0 (which is the current default value for limit). I'm not sure that limit of 0 is really valid.

Reply:
with the above changes a limit of `0` isn't possible anymore